# Assessment Schedule


| **Assessment** | **Subject** | **Weight** | **Due Date** | 
|----------------|-------------|------------|--------------|
| Assessment 1 | Command-line workshop | 9% | Friday 12th August |
| Assessment 2 | Scripting + Alignment | 9% |  |
| Assessment 3 | Sequencing | 9% |  |
| Assessment 4 | R programming + Downstream Analysis | 9% |  |
| Assessment 5 | Metagenomics | 9% |  | 
| Project | Complete Dataset | 15% |  |
| Exam | | 40% | | 

### Assessment Checklist

- Finish Answering the question
- Naming conventions for Assessments
	- Assessment coversheets are required for each assessment with a signed commitment that you have read the University's policy on assessment plagiarism. Any indication that material has been copied from student to student will result in a zero for that assignment 
	- If you are handing in written/paper assessment, each sheet must also have your name and student number (header & footer would be useful if you are using word format)
- Communication with assessors
	- Emails must contain subject information with your information. e.g. "BIO7005: Jimmy Breen Assignment1"
- Assessments are due at the Friday tutorial 10am
	
#### Major Project (15%)

In this course, the following next-generation sequencing (NGS) datasets/protocols will be in detail:

1. Whole genome sequencing/Resequencing  
2. Transcriptome Sequencing (RNAseq)
3. DNA Methylation/Bisulfite Sequencing
4. Enrichment/Capture sequencing (Methyl-capture, ChIPseq, RIPseq)
5. Metagenomics/Microbial profiling

Each of these NGS approaches uses similar programs and analysis approaches, such as quality control (quality and sequencing adapter trimming), genome alignment, and downstream visualisation and statistical methods. For the major project, you will take a published dataset and complete all the analysis tasks (from raw data to final results) during the course. 

You have the freedom to choose any dataset you would like before we start week 5's practical (NGS data practical), or choose from four datasets that we have chosen. You will be asked assessment questions on your data during practical assessments throughout the semester, and a final report will be due the week before the end of semester (Week 12)


#### Practical Assignment 1 - Introduction to command-line (9%)

Week 1-2

__Question 1__

Submit a simplified version of your terminal history (copy-pasting will suffice) for the following commands:

__Question 2__

At the start of the practical you told that the following programming conventions should be followed at all time:
	- Do not use spaces in filenames
	- Extensively comment your code
	- Robust error handling
	- Report version numbers when using programs
	- Use sensible variable names
	- File checks: Checking if your code is valid and runs successfully
	
Explain the importance of these programming conventions 

For additional information, please look at the following websites:

[How To Ask Questions The Smart Way](http://www.catb.org/esr/faqs/smart-questions.html)

[How to write a good bug report](https://musescore.org/en/developers-handbook/how-write-good-bug-report-step-step-instructions)

#### Practical Assignment 2 - Introduction to command-line scripting (9%)

Week 3-4

__Question 1__



__Question 2__


#### Practical Assignment 3 - Next-Generation Sequencing (9%)

Week 5-6

__Question 1__



__Question 2__


#### Practical Assignment 4 - R programming (9%)

Week 7-8

__Question 1__



__Question 2__


#### Practical Assignment 5 - Systems Biology (9%)

Week 9

__Question 1__



__Question 2__


#### Practical Assignment 6 - Metagenomics (9%)

Week 10-11

__Question 1__



__Question 2__
